### PR TITLE
Don't generate unnecessary fresh symbols for the GOTO trace

### PR DESCRIPTION
**Pull request by tautschnig**
We can safely record the values of expressions by adding `expr == expr`
as constraints in order to be able to fetch and display them in the GOTO
trace. This was already being done for declarations. Introducing new
symbols just adds unnecessary variables to the formula.

When running on various proofs done for AWS open-source projects, this
changes the performance as follows: with CaDiCaL as back-end, the total
solver time for the hardest 46 proofs changes from 26546.5 to 26779.7
seconds (233.2 seconds slow-down); with Minisat, however, the hardest 49
proofs take 28420.4 instead of 32387.2 seconds (3966.8 seconds
speed-up). Across these benchmarks, 1.7% of variables and 0.6% of
clauses are removed.

develop-vs-no-fresh

- [x] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [x] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Comment by martin-cs** [comment]
I love the performance graphs but please can we have the diagonal line so it is easy to see better / worse.

File: regression/cbmc-incr-smt2/pointers-conversions/pointer_to_int.desc [hunk 1]
```diff
@@ -1,4 +1,4 @@
-CORE
+KNOWNBUG
```
Comment (martin-cs) [inline regression/cbmc-incr-smt2/pointers-conversions/pointer_to_int.desc:1]: Why is this broken?

**Review by martin-cs — APPROVED** [review]
It is a little hard to follow how the changes implement what is described in the PR.

File: regression/cbmc-incr-smt2/pointers-conversions/pointer_to_int.desc (hunk 1 above)
Comment (tautschnig) [reply regression/cbmc-incr-smt2/pointers-conversions/pointer_to_int.desc:1]: Additional work is required in the new SMT back-end to make `get_value` work for this will now see arbitrary expressions rather than just symbols. The other back-ends already handle this case properly.

Comment (martin-cs) [reply regression/cbmc-incr-smt2/pointers-conversions/pointer_to_int.desc:1]: Fair.

**Comment by kroening** [comment]
This is not a good idea.

I'd first like to understand how the speedup works. And then I'd like to push that change into the decision procedure. The user of the decision procedure (here symex) shouldn't have to care about encoding tricks that are this low level.
